client/shaders/nodes_shader/opengl_fragment.glsl

```diff
@@ -189,6 +189,11 @@ void main(void)
 	}
 #endif
 	vec4 base = texture2D(baseTexture, uv).rgba;
+	// If alpha is zero, we can just discard the pixel. This fixes transparency
+	// on GPUs like GC7000L, where GL_ALPHA_TEST is not implemented in mesa.
+	if (base.a == 0.0) {
+		discard;
```

Comment:
The performance of something like this is actually horrible, especially if you're doing it on all objects unconditionally. You lose pretty much all internal hardware optimizations by making the fragment's existence dependent on its execution's result.

Reply:
Your criticism is strongly worded, but I believe it is incorrect. The accepted answer on the stackoverflow question you are quoting states "Note that you would get the same issue from turning on alpha test." Which is what minetest / irrlicht is currently using (GL_ALPHA_TEST). The optimization you are probably referring to is Early-Z Reject, which has to be automatically turned off regardless of the strategy that is used to reveal pixels behind a transparent pixel in a triangle. How do you think it would work otherwise?

If you had a frame with no objects featuring transparency or blending, then this if-statement might make performance worse. You could use a different shader for 1. objects with transparency or blending and 2. fully solid objects but this would increase code complexity.

Reply:
I was thinking about this more, and in general, to optimize this in minetest, shouldn't there be two passes:
1. Paint all solid objects, leveraging early-z reject (no discard, no alpha test)
2. Paint all objects with transparency (with a shader that uses discard, or with alpha test or blending)

Reply:
> I was thinking about this more, and in general, to optimize this in minetest, shouldn't there be two passes:
> 
> 1. Paint all solid objects, leveraging early-z reject (no discard, no alpha test)
> 2. Paint all objects with transparency (with a shader that uses discard, or with alpha test or blending)

In all engines I've worked with non-opaques come after opaques. Even if there are gpu techniques that allow mixing of opaques and (multiple) translucencies and/or punch-throughs, doing those in one pass is always sub-optimal, for the rudimentary reason that you don't want to unnecessary blend or fragment-test opaque primitives -- that's BW waste normally paid in fillrate.

Re Early-(coarse/hierarchical)-Z (where fragments, quads or even entire prims are z-culled before fragment-shader time) -- the discard / punch-through is incompatible not because of the rejection aspect, but because of the acceptance aspect: while a punched-through primitive *can* be discarded early-on, it cannot perform a depth update early-on -- batch depth is not known until the shader of the entire batch gets executed. The situation is even worse with dept-update primitives (which compute their gl_FragDepth, rather than use the rasterizer-supplied value) -- they cannot even be z-culled until every relevant fragment has its depth computed; for them any form of Early-Z is non-applicable, down to a fragment level. Naturally, if the GPU treats fragment-discard the same way as fragment-depth-update, no Early-Z of any form will be applicable.

Reply:
Yes, I understand that. Thank you for elaborating on this. My first point was that GL_ALPHA_TEST, which is currently used, has the same effect as discard in terms of Early-Z, do you agree?

Reply:
If Minetest already draws everything opaque with alpha test, then that's a serious blunder. Would explain the crap fps in more busy places, like jungles.

Reply:
> Yes, I understand that. Thank you for elaborating on this. My first point was that GL_ALPHA_TEST, which is currently used, has the same effect as discard in terms of Early-Z, do you agree?

I've seen drivers where *any* kind of fragment-test is implemented via implicit discard -- so, yes, I do agree. That was the point of my elaboration -- your suggestion is viable, and it would make sense if minetest adopted such a differentiation.